src/cpu_shader/euler.rs

```diff
@@ -0,0 +1,277 @@
+        */
+        ctr + 1.55 * aerr + halo_symm + halo_asymm
+    }
+}
+
+impl EulerParams {
+    pub fn from_angles(th0: f32, th1: f32) -> EulerParams {
+        let k0 = th0 + th1;
+        let dth = th1 - th0;
+        let d2 = dth * dth;
+        let k2 = k0 * k0;
+        let mut a = 6.0;
+        a -= d2 * (1. / 70.);
+        a -= (d2 * d2) * (1. / 10780.);
+        a += (d2 * d2 * d2) * 2.769178184818219e-07;
+        let b = -0.1 + d2 * (1. / 4200.) + d2 * d2 * 1.6959677820260655e-05;
+        let c = -1. / 1400. + d2 * 6.84915970574303e-05 - k2 * 7.936475029053326e-06;
+        a += (b + c * k2) * k2;
+        let k1 = dth * a;
+
+        // calculation of chord
+        let mut ch = 1.0;
+        ch -= d2 * (1. / 40.);
+        ch += (d2 * d2) * 0.00034226190482569864;
+        ch -= (d2 * d2 * d2) * 1.9349474568904524e-06;
+        let b = -1. / 24. + d2 * 0.0024702380951963226 - d2 * d2 * 3.7297408997537985e-05;
+        let c = 1. / 1920. - d2 * 4.87350869747975e-05 - k2 * 3.1001936068463107e-06;
+        ch += (b + c * k2) * k2;
+        EulerParams {
+            th0,
+            th1,
+            k0,
+            k1,
+            ch,
+        }
+    }
+
+    pub fn eval_th(&self, t: f32) -> f32 {
+        (self.k0 + 0.5 * self.k1 * (t - 1.0)) * t - self.th0
+    }
+
+    /// Evaluate the curve at the given parameter.
+    ///
+    /// The parameter is in the range 0..1, and the result goes from (0, 0) to (1, 0).
+    fn eval(&self, t: f32) -> Vec2 {
+        let thm = self.eval_th(t * 0.5);
+        let k0 = self.k0;
+        let k1 = self.k1;
+        let (u, v) = integ_euler_10((k0 + k1 * (0.5 * t - 0.5)) * t, k1 * t * t);
+        let s = t / self.ch * thm.sin();
+        let c = t / self.ch * thm.cos();
+        let x = u * c - v * s;
+        let y = -v * c - u * s;
+        Vec2::new(x, y)
+    }
+
+    fn eval_with_offset(&self, t: f32, offset: f32) -> Vec2 {
+        let th = self.eval_th(t);
+        let v = Vec2::new(offset * th.sin(), offset * th.cos());
+        self.eval(t) + v
+    }
+}
+
+impl EulerSeg {
+    pub fn from_params(p0: Vec2, p1: Vec2, params: EulerParams) -> Self {
+        EulerSeg { p0, p1, params }
+    }
+
+    #[allow(unused)]
+    pub fn eval(&self, t: f32) -> Vec2 {
+        let Vec2 { x, y } = self.params.eval(t);
+        let chord = self.p1 - self.p0;
+        Vec2::new(
+            self.p0.x + chord.x * x - chord.y * y,
+            self.p0.y + chord.x * y + chord.y * x,
+        )
+    }
+
+    pub fn eval_with_offset(&self, t: f32, offset: f32) -> Vec2 {
+        let chord = self.p1 - self.p0;
+        let scaled = offset / chord.length();
+        let Vec2 { x, y } = self.params.eval_with_offset(t, scaled);
+        Vec2::new(
+            self.p0.x + chord.x * x - chord.y * y,
+            self.p0.y + chord.x * y + chord.y * x,
+        )
+    }
+}
+
+/// Integrate Euler spiral.
+///
+/// TODO: investigate needed accuracy. We might be able to get away
+/// with 8th order.
```

Comment:
Eventually we should have a wiki page (as well as the definitive paper) that explain the formulations here. Maybe add a disclaimer for now that a lot of the magic numbers here were derived by hand and needs to be documented?

Reply:
I wrote a paragraph explaining the tradeoff. It's also something that can potentially be explained in the paper.